usecase/github_user.go

```diff
@@ -0,0 +1,25 @@
+//Package usecase processes the information from repository, making any needed calculations, data transformation, etc.
+//If it recieved data it's from controllers, coming from delivery layer
+package usecase
+
+import (
+	"lolidelgado/github-users/models"
+	"lolidelgado/github-users/repository"
+)
+
+type GithubUserUseCase struct {
+	githubUserRepo *repository.GithubUser
+}
+
+func NewGithubUser(githubUserRepo *repository.GithubUser) *GithubUserUseCase {
+	return &GithubUserUseCase{
+		githubUserRepo,
+	}
+}
+
+func (g *GithubUserUseCase) FetchAll() ([]models.GithubUser, error) {
+	return g.githubUserRepo.FetchAll()
+}
```

Comment:
Something important in Go at the time of determining the root cause of an error is to wrap errros.

Let's say you have more logic here, and you simply decide to return what you get from the repository, and also let's imagine you need to determine the reason of this error. You won't know at glance if the repo package, or the usecase package, or any other package was the pace  where the error was originated.

By wrapping the error in each layer you are adding useful information that will help debugging tasks.

To wrap errors, there's a method named `Errorf` in the `fmt`pakage.

```suggestion
func (g *GithubUserUseCase) FetchAll() ([]models.GithubUser, error) {
	githubUsers, err := g.githubUserRepo.FetchAll()
	if err != nil {
		return nil, fmt.Errorf("Error fetching github users: %w", err)
	}

	return githubUsers, nil
}

```

Reply:
Is it better if I handle the errors from the bottom, indicating errors and reasons from the repository, I am going for that approach but please let me know if there is another way that is more suggested as best practice.

Reply:
Well, the most common way I've seen in my experience is to back-propagate the error in each layer and adding information in each one of the layers; there is not a best practice that reigns above all, but reading [this](https://earthly.dev/blog/golang-errors/ ) might come handy to take a decision.

- You can return an error created on the fly, just as the suggestion I added.
- You can initialize it as a package-level error variable and use it multiple times.
- You can define your own errors package, with error definitions, methods, and functions for the whole project.

Any of the choice you take, remember that you can also define your own errors; `error` is an interface that consists on only an `Error() string` method.

Choose the approach that better suits your needs; remember that part of the deliverable is to include unit tests for your code; maybe an already defined error works better, but the approach is according to how exactly you want to organize your project

And remember, it's always a good idea to wrap errors.